### PR TITLE
Template transaction database

### PR DESCRIPTION
**Pull request by zeptin**
Fix: #100 

Note that the template transaction class is currently just a stub, another issue will be created regarding specifying its contents.

It will be necessary to define a canonical hash to identify the template transaction within the database. Possibly this will be the hash of the deterministically generated transaction without any signatures in it.

The `GetAsync` method currently causes a deserialisation error within DBreeze. This needs to be investigated further.

### CODE REVIEW

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.DBreeze = new DBreezeEngine(folder);
+            this.network = network;
+            this.dateTimeProvider = dateTimeProvider;
+        }
+
+        /// <summary>Performs any needed initialisation for the database.</summary>
+        public virtual Task InitializeAsync()
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: I think they have started using Fody and writing interceptors for this kind of entry and exit logs in the full node, I'll ask for a new package soon and we can do it the same way :)

Comment (zeptin) [reply src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: Agreed

**Comment by zeptin** [comment]
Note that GetAsync is failing as the DBreezeSerializer does not know how to handle TemplateTransaction instances.

See: [#2331](https://github.com/stratisproject/StratisBitcoinFullNode/issues/2331)

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransaction.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using NBitcoin;
+
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransaction.cs]: Not sure you need a backing property in there? Maybe I am wrong :/ is it something DBreeze specific ?

Comment (monsieurleberre) [reply src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransaction.cs]: Also wondering if we don't need some form of status stored in the database, like
pending-signature, in-mempool, on-chain, matured-on-chain, rejected, etc.
Or do you plan to put that in another object like "transfer" which would reference these templates ?

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+        /// <summary>
+        /// Determine if a template transaction already exists in the database.
+        /// </summary>
+        /// <param name="hash">The hash of the template transaction.</param>
+        /// <returns><c>true</c> if the template transaction can be found in the database, otherwise return <c>false</c>.</returns>
+        Task<bool> ExistAsync(uint256 hash);
+    }
+
+    public class TemplateTransactionStore : ITemplateTransactionStore
```
Comment (monsieurleberre) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: If I understand that correctly, this is the store where we plan to have all the different crosschain transfers and their different statuses ?

**Comment by quantumagi** [comment]
From reading the spec it looks like we should be needing the following storage:

```
Chain A - Send Store
----------------------
- Deposit Transaction ID
- Deposit Block Height
- Deposit Target Address
- Deposit Amount
- Chain B Acknowledged

Chain B - Receive Store
-------------------------
- Deposit Transaction ID
- Deposit Block Height
- Deposit Target Address
- Deposit Amount
- Partial Transaction Hex
- Partial Transaction Hash
- Transaction Block Height
- Signatures
- Status
  - "Deposit Received"
  - "Partial Transaction Created" (Depends on UTXO availability)
  - "Fully signed"
  - "Broadcasted"
  - "Seen in Block"
  - "Complete" (Reorg depth reached)
```

**Comment by zeptin** [comment]
@quantumagi That looks good. Bearing in mind the two chains are symmetric, you will have both deposit and withdrawal transactions to deal with. I think the same store design can be used for both, with just a flag to indicate direction perhaps.

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        /// <summary>Performs any needed initialisation for the database.</summary>
+        public virtual Task InitializeAsync()
+        {
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                // Currently don't do anything on startup
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: Doesn't anything need to be initialised here ? current tip ? (I'm not fully sure what the template transaction does, but the common store pattern in the full node was to keep the current block hash height - maybe it's not need here).

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<bool> ExistAsync(uint256 hash)
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: Mentioned above maybe share the function with `GetAsync()`.

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+            return task;
+        }
+
+        protected virtual void OnInsertTemplateTransaction(DBreeze.Transactions.Transaction dbreezeTransaction, TemplateTransaction templateTransaction)
+        {
+            // If the template is already in store don't write it again.
+            Row<byte[], TemplateTransaction> templateRow = dbreezeTransaction.Select<byte[], TemplateTransaction>(TemplateTransactionTableName, templateTransaction.Hash.ToBytes());
+
+            if (!templateRow.Exists)
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: If keeping `ExistAsync()` then could that here instead.

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        protected virtual void OnInsertTemplateTransaction(DBreeze.Transactions.Transaction dbreezeTransaction, TemplateTransaction templateTransaction)
+        {
+            // If the template is already in store don't write it again.
+            Row<byte[], TemplateTransaction> templateRow = dbreezeTransaction.Select<byte[], TemplateTransaction>(TemplateTransactionTableName, templateTransaction.Hash.ToBytes());
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: If keeping `ExistAsync()` then could that here instead.

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+            if (!templateRow.Exists)
+            {
+                dbreezeTransaction.Insert<byte[], TemplateTransaction>(TemplateTransactionTableName, templateTransaction.Hash.ToBytes(), templateTransaction);
+            }
+        }
+
+        protected virtual void OnDeleteTemplateTransaction(DBreeze.Transactions.Transaction dbreezeTransaction, uint256 hash)
+        {
+            dbreezeTransaction.RemoveKey<byte[]>(TemplateTransactionTableName, hash.ToBytes());
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: Could remove this method and move this line up to line 183.

File: src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                    transaction.Commit();
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        protected virtual void OnInsertTemplateTransaction(DBreeze.Transactions.Transaction dbreezeTransaction, TemplateTransaction templateTransaction)
```
Comment (ferdeen) [inline src/Stratis.FederatedPeg.Features.FederationGateway/TemplateTransactionStore.cs]: 'On' suggests an event or signal has happened, maybe change to `InsertTemplateTransaction`.

**Comment by quantumagi** [comment]
After discussing with Matt we don't need a "send store".

So this means the above changes as follows:

Chain B - Receive Store
-------------------------
- Deposit Transaction ID
- Deposit Block Height
- Deposit Target Address
- Deposit Amount
- Partial Transaction Hex
- Partial Transaction Hash
- Transaction Block Height
- Signatures
- Status
  - "Deposit Received"
  - "Partial Transaction Created" (Depends on UTXO availability)
  - "Fully signed"
  - "Broadcasted"
  - "Seen in Block"
  - "Complete" (Reorg depth reached)

Also node B needs to keep track of:
- the last chain B block (the "store tip") that information is included for in the above store.
- the last chain A block that that information is included for in the above store.

We can track that in a separate table on DBreeze (e.g. "Common").

**Comment by monsieurleberre** [comment]
@quantumagi yes I agree, maybe two comments
- adding transaction block hash (from the target chain, or B chain) just in case that chain splits for a minute
- in the statuses, there should be a "rejected" maybe in case you are asked to sign a transaction that you can't back with a deposit on the chain A

**Comment by quantumagi** [comment]
Merged this but work is continuing on this.

**Pull request merged**